src/basic/process-util.c

```diff
@@ -312,19 +312,18 @@ int rename_process(const char name[]) {
         /* Third step, completely replace the argv[] array the kernel maintains for us. This requires privileges, but
          * has the advantage that the argv[] array is exactly what we want it to be, and not filled up with zeros at
          * the end. This is the best option for changing /proc/self/cmdline. */
-        if (mm_size < l+1) {
+
+        /* Let's not bother with this if we don't have euid == 0. Strictly speaking if people do weird stuff with
+         * capabilities this could work even for euid != 0, but our own code generally doesn't do that, hence let's
+         * use this as quick bypass check, to avoid calling mmap() if PR_SET_MM_ARG_{START,END} fails with EPERM later
+         * on anyway. After all geteuid() is dead cheap to call, but mmap() is not. */
+        if (geteuid() != 0) {
```

Comment:
This check doesn't seem to be sufficient. A command line still can be corrupted if the process calling `rename_process` doesn't have the `CAP_SYS_RESOURCE` capability. The following patch can be used to trigger the corruption:
```diff
diff --git a/src/test/test-process-util.c b/src/test/test-process-util.c
index e7c9766..6319808 100644
--- a/src/test/test-process-util.c
+++ b/src/test/test-process-util.c
@@ -32,6 +32,7 @@

 #include "alloc-util.h"
 #include "architecture.h"
+#include "capability-util.h"
 #include "fd-util.h"
 #include "log.h"
 #include "macro.h"
@@ -423,6 +424,7 @@ static void test_rename_process_multi(void) {

         /* child */
         test_rename_process_now("one", 1);
+        drop_capability(CAP_SYS_RESOURCE);
         test_rename_process_now("more", 0); /* longer than "one", hence truncated */
         setresuid(99, 99, 99);
         test_rename_process_now("time!", 0);
```
I'm not sure what to do about it. The function might check that the process has the `CAP_SYS_RESOURCE` capability. Unfortunately, that won't work inside a user namespace.